grpc/src/main/java/com/linecorp/armeria/server/grpc/UnframedGrpcErrorHandlerBuilder.java

```diff
@@ -0,0 +1,245 @@
+     */
+    public UnframedGrpcErrorHandlerBuilder jsonMarshaller(MessageMarshaller jsonMarshaller) {
+        requireNonNull(jsonMarshaller, "jsonMarshaller");
+        checkState(marshalledMessages == null && marshalledMessageTypes == null,
+                   "Cannot set a custom JSON marshaller because one or more Message or MessageType instances" +
+                   " have already been registered. To set a custom marshaller, ensure that no Message or" +
+                   " MessageType registrations have been made before calling this method."
+        );
+        this.jsonMarshaller = jsonMarshaller;
+        return this;
+    }
+
+    /**
+     * Specifies the status mapping function to be used by the error handler.
+     *
+     * <p>This function determines how gRPC statuses are mapped to HTTP statuses
+     * in the error response.
+     *
+     * @param statusMappingFunction The status mapping function
+     */
+    public UnframedGrpcErrorHandlerBuilder statusMappingFunction(
+            UnframedGrpcStatusMappingFunction statusMappingFunction) {
+        this.statusMappingFunction = requireNonNull(statusMappingFunction, "statusMappingFunction");
+        return this;
+    }
+
+    /**
+     * Specifies the response types that the error handler will support.
+     *
+     * <p>This method allows specifying one or more response types (e.g., JSON, PLAINTEXT)
+     * that the error handler can produce. If nothing is specified or multiple types are specified, the actual
+     * response type is determined by the response's content type.
+     *
+     * @param responseTypes The response types to support
+     */
+    public UnframedGrpcErrorHandlerBuilder responseTypes(UnframedGrpcErrorResponseType... responseTypes) {
+        requireNonNull(responseTypes, "responseTypes");
+        if (this.responseTypes == null) {
+            this.responseTypes = EnumSet.noneOf(UnframedGrpcErrorResponseType.class);
+        }
+        this.responseTypes.addAll(ImmutableSet.copyOf(responseTypes));
+        return this;
+    }
+
+    /**
+     * Registers custom messages to be marshalled by the error handler.
+     *
+     * <p>This method registers specific message instances for custom error responses.
+     * If a custom JSON marshaller has already been set, calling this method will
+     * result in an {@link IllegalStateException}.
+     *
+     * @param messages The message instances to register
+     */
+    public UnframedGrpcErrorHandlerBuilder registerMarshalledMessages(Message... messages) {
+        requireNonNull(messages, "messages");
+        checkState(jsonMarshaller == null,
+                   "Cannot register custom messages because a custom JSON marshaller has already been set. " +
+                   "Use the custom marshaller to register custom messages.");
+        if (marshalledMessages == null) {
+            marshalledMessages = new ArrayList<>();
+        }
+        marshalledMessages.addAll(Arrays.asList(messages));
```

Comment:
Could be simplified into:
```suggestion
        Collections.addAll(marshalledMessages, messages);
```